README.md

This extension was written and published by Adrian Lungu. It was on Github at some point.

It was removed from Chrome Web Store at some point. :(

I (paulirish) recovered it and put it on Github. Because I like it.



### To install

Install as an unpacked extension. To do that:

* Download this repo or clone it
* Open chrome://extensions
* Enable "Developer mode" at the top of the page
* Click "Load unpacked" and find the folder of your checkout (where `manifest.json` is). Select it.
* Ta-da! It's installed.

Once the window is open, click the hamburger menu for a few preferences. There's also a bunch of options under "More options".

### Screenshot

<img src="https://user-images.githubusercontent.com/39191/52365295-068e5780-29fc-11e9-9329-88a3bcf719c2.png" width=250>

CHANGELOG
V 2.1.0 (UPCOMING)
    - TODO: Allow to move tabs by dragging to a new position
    - TODO: Close when all Chrome windows are closed

V 2.0.2 (CURRENT)
    - Removed some permissions which were not needed

V 2.0.1
    - Fixed a bug where closing a tab would click it before closing it making a tab next to it active

V 2.0.0
    - Minimize and restore along with the main Chrome window which opened the extension
    - Side tabs window adapts position based on Chrome main window
    - Move some settings to a Settings button inside Side Tabs Window
    - Ability to show pinned tabs
    - Filter tabs
    - Memorize Main Window position to restore it once Side Tabs are closed


app/chrome/background.js
(function () {

    var idSideTabsWindow;
    var idCurrentChromeWindow;
    var isLeftSide = true;
    var width = 250;

    function initSideTabs() {

        if (width < 30 || width > screen.width - 30) {
            width = 250;
        }

        chrome.windows.update(idCurrentChromeWindow,
            {
                'left': isLeftSide ? width : 0,
                'width': screen.width - width
            }
        );

        chrome.windows.create({
            'url': chrome.extension.getURL('app/index.html#?idChromeWindow=' + idCurrentChromeWindow + '&screenWidth=' + screen.width),
            'left': isLeftSide ? 0 : screen.width - width,
            'top': 0,
            'width': width,
            'height': screen.height,
            'type': 'popup'
        }, function (window) {
            idSideTabsWindow = window.id;
        });

    }

    chrome.browserAction.onClicked.addListener(function () {

        chrome.windows.getCurrent({}, function (window) {

            idCurrentChromeWindow = window.id;

            chrome.storage.sync.get(['optionsSettings', 'sideTabsSettings'], function (items) {

                if (!items.optionsSettings) {
                    items.optionsSettings = {
                        openSide: 'left',
                        backgroundColor: '#D4D4D4',
                        singleInstance: true,
                        autoAdjustWidth: true
                    }
                } else {
                    items.optionsSettings = JSON.parse(items.optionsSettings);
                }

                if (!items.sideTabsSettings) {
                    items.sideTabsSettings = {
                        tabWidth: 250,
                        showTabsForSelectedWindow: false,
                        showTabsGroupedPerWindow: false,
                        showPinnedTabs: false
                    }
                } else {
                    items.sideTabsSettings = JSON.parse(items.sideTabsSettings);
                }

                isLeftSide = items.optionsSettings.openSide === 'left';
                width = items.sideTabsSettings.tabWidth;

                if (items.optionsSettings.singleInstance === true) {
                    if (!idSideTabsWindow) {
                        initSideTabs();
                    }
                    else {
                        chrome.windows.get(idSideTabsWindow, function (window) {
                            if (chrome.runtime.lastError) {
                                initSideTabs();
                            } else if (!window) {
                                initSideTabs();
                            }
                            else {
                                chrome.windows.update(idSideTabsWindow, {'focused': true});
                            }
                        })
                    }
                } else {
                    initSideTabs();
                }
            })
        });

    });

})();

app/chrome/options/options.js
/**
 * Created by adrianlungu on 08/11/14.
 */

(function () {

    "use strict";

    var tab = document.querySelector('.tab');

    var tabSettings = {

        height: document.querySelector('#tabHeight'),
        fontSize: document.querySelector('#tabFontSize'),
        textColor: document.querySelector('#tabTextColor'),

        backgroundColor: {
            top: document.querySelector('#tabBackgroundTopColor'),
            bottom: document.querySelector('#tabBackgroundBottomColor')
        },

        hoverColor: {
            top: document.querySelector('#tabBackgroundHoverTopColor'),
            bottom: document.querySelector('#tabBackgroundHoverBottomColor')
        },

        highlightColor: document.querySelector('#tabBackgroundHighlightColor'),
        borderColor: document.querySelector('#tabBorderColor')

    };

    var settings = {

        openSide: document.querySelector('#tabOpenSide'),
        backgroundColor: document.querySelector('#backgroundColor'),
        singleInstance: document.querySelector('#singleInstance'),
        autoAdjustWidth: document.querySelector('#autoAdjustWidth'),
        autoMinimize: document.querySelector('#autoMinimize')

    };

    var version = document.querySelector('#version');
    var save = document.querySelector('#save');

    var tabFunctions = {
        mouseEnter: function () {
            tab.style.background = 'linear-gradient(' + tabSettings.hoverColor.top.value + ', '
                + tabSettings.hoverColor.bottom.value + ')';
        },
        mouseLeave: function () {
            tab.style.background = 'linear-gradient(' + tabSettings.backgroundColor.top.value + ', '
                + tabSettings.backgroundColor.bottom.value + ')';
        },
        mouseDown: function () {
            tab.style.background = tabSettings.highlightColor.value;
        }
    };

    function saveOptions() {

        chrome.storage.sync.set({

                tab: JSON.stringify({
                    height: parseInt(tabSettings.height.value),
                    fontSize: parseInt(tabSettings.fontSize.value),
                    textColor: tabSettings.textColor.value,

                    backgroundColor: {
                        top: tabSettings.backgroundColor.top.value,
                        bottom: tabSettings.backgroundColor.bottom.value
                    },

                    hoverColor: {
                        top: tabSettings.hoverColor.top.value,
                        bottom: tabSettings.hoverColor.bottom.value
                    },

                    highlightColor: tabSettings.highlightColor.value,
                    borderColor: tabSettings.borderColor.value
                }),

                optionsSettings: JSON.stringify({
                    openSide: settings.openSide.value,
                    backgroundColor: settings.backgroundColor.value,
                    singleInstance: settings.singleInstance.checked,
                    autoAdjustWidth: settings.autoAdjustWidth.checked,
                    autoMinimize: settings.autoMinimize.checked
                })

            }, function () {
                save.innerHTML = 'Options Saved!';

                setTimeout(function () {
                    save.innerHTML = 'Save';
                }, 1000)
            }
        )
    }

    function getUserDefinedSettings() {

        chrome.storage.sync.get(['tab', 'optionsSettings'], function (items) {

            if (!items.tab) {
                items.tab = {
                    height: 25,
                    fontSize: 11,
                    textColor: '#000000',

                    backgroundColor: {
                        top: '#DBD7D9',
                        bottom: '#C4C3C5'
                    },

                    hoverColor: {
                        top: '#F5F5F5',
                        bottom: '#E6E6E6'
                    },

                    highlightColor: '#F5F5F5',
                    borderColor: '#A9A9A9'
                }
            } else {
                items.tab = JSON.parse(items.tab);
            }

            if (!items.optionsSettings) {
                items.settings = {
                    openSide: 'left',
                    backgroundColor: '#D4D4D4',
                    singleInstance: true,
                    autoAdjustWidth: true,
                    autoMinimize: true
                }
            } else {
                items.settings = JSON.parse(items.optionsSettings);
            }

            tabSettings.height.value = items.tab.height;
            tabSettings.fontSize.value = items.tab.fontSize;
            tabSettings.textColor.value = items.tab.textColor;

            tabSettings.backgroundColor.top.value = items.tab.backgroundColor.top;
            tabSettings.backgroundColor.bottom.value = items.tab.backgroundColor.bottom;

            tabSettings.hoverColor.top.value = items.tab.hoverColor.top;
            tabSettings.hoverColor.bottom.value = items.tab.hoverColor.bottom;

            tabSettings.highlightColor.value = items.tab.highlightColor;
            tabSettings.borderColor.value = items.tab.borderColor;

            settings.openSide.value = items.settings.openSide;
            settings.backgroundColor.value = items.settings.backgroundColor;
            settings.singleInstance.checked = items.settings.singleInstance;
            settings.autoAdjustWidth.checked = items.settings.autoAdjustWidth;
            settings.autoMinimize.checked = items.settings.autoMinimize;

            setTabStyle();

        })

    }

    function setTabStyle() {

        tab.style.color = tabSettings.textColor.value;
        tab.style.background = 'linear-gradient(' + tabSettings.backgroundColor.top.value + ', '
            + tabSettings.backgroundColor.bottom.value + ')';
        tab.style.borderColor = tabSettings.borderColor.value;
        tab.style.height = tabSettings.height.value + 'px';
        tab.style.fontSize = tabSettings.fontSize.value + 'px';

    }

    function init() {

        getUserDefinedSettings();

        version.innerHTML = 'Version ' + chrome.runtime.getManifest().version;

        save.addEventListener("click", saveOptions);

        tabSettings.height.addEventListener("input", setTabStyle);
        tabSettings.fontSize.addEventListener("input", setTabStyle);
        tabSettings.textColor.addEventListener("input", setTabStyle);

        tabSettings.backgroundColor.top.addEventListener("input", setTabStyle);
        tabSettings.backgroundColor.bottom.addEventListener("input", setTabStyle);

        tabSettings.hoverColor.top.addEventListener("input", setTabStyle);
        tabSettings.hoverColor.bottom.addEventListener("input", setTabStyle);

        tabSettings.highlightColor.addEventListener("input", setTabStyle);
        tabSettings.borderColor.addEventListener("input", setTabStyle);

        tab.addEventListener('mouseenter', tabFunctions.mouseEnter);
        tab.addEventListener('mouseleave', tabFunctions.mouseLeave);

        tab.addEventListener('mousedown', tabFunctions.mouseDown);
        tab.addEventListener('mouseup', tabFunctions.mouseEnter);

    }

    init();

})();


app/core/directives/tab/tab.js
(function () {
    'use strict';

    angular
        .module('app.core')
        .directive('tab', tab);

    tab.$inject = [];

    /* @ngInject */
    function tab() {
        var directive = {
            link: link,
            restrict: 'AE',
            scope: {
                tab: '=',
                window: '=',
                idLastChromeWindow: '=',

                onClick: '&',
                onClose: '&'
            },
            templateUrl: chrome.extension.getURL('app/core/directives/tab/tab.html')
        };
        return directive;

        function link(scope, element, attrs) {

            scope.checkHighlight = function (newValue) {
                scope.tabClass = "";

                if (scope.tab.highlighted == true && scope.idLastChromeWindow == scope.tab.windowId) {
                    if (scope.tabClass.indexOf("highlighted") == -1) {
                        scope.tabClass += " highlighted ";
                    }
                } else {
                    scope.tabClass.replace(" highlighted ", "");
                }
            };

            scope.$watch('idLastChromeWindow', function (newValue, oldValue) {
                scope.checkHighlight(newValue)
            }, true);

            scope.$watch('tab.highlighted', function (newValue, oldValue) {
                scope.checkHighlight(newValue)
            }, true);

            var close = angular.element(element[0].querySelector('.close'))[0];
            close.addEventListener('mousedown', function(e) {
                e.stopPropagation();
                scope.onClose({tab: scope.tab});
            });

            var tab = angular.element(element[0].querySelector('.tab'))[0];
            tab.addEventListener('mousedown', function (e) {

                e.stopPropagation();

                var buttonCode = e.button;

                switch(buttonCode) {
                    case 0:
                        scope.onClick({tab: scope.tab});
                        break;

                    case 1:
                        scope.onClose({tab: scope.tab});
                        break;
                }

            });

        }

    }

})();

app/core/core.controller.js
(function () {
    "use strict";

    angular
        .module('app.core')
        .controller('coreController', coreController);

    coreController.$inject = ['$scope', '$interval', '$location'];

    /* @ngInject */
    function coreController($scope, $interval, $location) {
        var vm = this;

        vm.data = {
            idSideTabsWindow: 0,
            idLastChromeWindow: 0,
            search: '',

            isNewTabFromSideTabs: false,

            windows: [],
            numberOfTabs: 0,
        };
        vm.optionsSettings = {
            openSide: 'left',
            backgroundColor: '',
            autoAdjustWidth: true
        };
        vm.settings = {
            tabWidth: 250,
            showSettings: false,
            showTabsForSelectedWindow: false,
            showTabsGroupedPerWindow: false,
            showPinnedTabs: false,
            showSearch: false
        };
        vm.mainChromeWindow = {
            id: 0,
            left: 0,
            top: 0,
            width: 0,
            height: 0
        };

        vm.init = init;
        vm.initSideTabs = initSideTabs;
        vm.initChromeEvents = initChromeEvents;

        vm.initPreferences = initPreferences;
        vm.savePreferences = savePreferences;

        vm.clickCreateNewTab = clickCreateNewTab;
        vm.clickTab = clickTab;
        vm.clickCloseTab = clickCloseTab;
        vm.clickShowSettings = clickShowSettings;
        vm.clickShowSearch = clickShowSearch;
        vm.clickShowTabsForSelectedWindow = clickShowTabsForSelectedWindow;
        vm.clickShowTabsGroupedPerWindow = clickShowTabsGroupedPerWindow;
        vm.clickShowPinnedTabs = clickShowPinnedTabs;
        vm.clickMoreSettings = clickMoreSettings;

        vm.loadTabs = loadTabs;

        vm.isScrolledIntoView = isScrolledIntoView;
        vm.scrollTabIntoView = scrollTabIntoView;

        vm.checkChromeWindowPosition = checkChromeWindowPosition;
        vm.checkChromeWindowState = checkChromeWindowState;

        vm.onExit = onExit;

        // Chrome events

        vm.windowChanged = windowChanged;

        ////////////////

        function init() {

            vm.initPreferences();

        }

        function initSideTabs() {

            vm.mainChromeWindow.id = parseInt($location.search().idChromeWindow);
            vm.data.screenWidth = parseInt($location.search().screenWidth);
            vm.data.idLastChromeWindow = vm.mainChromeWindow.id;

            chrome.windows.getCurrent(function (window) {
                vm.data.idSideTabsWindow = window.id;
            });

            // INIT MAIN CHROME WINDOW
            chrome.windows.update(vm.mainChromeWindow.id,
                {
                    'left': vm.optionsSettings.openSide === 'left' ? vm.settings.tabWidth : 0,
                    'width': vm.data.screenWidth - vm.settings.tabWidth
                }
            );

            // Auto adjust width
            if (vm.optionsSettings.autoAdjustWidth) {
                $interval(vm.checkChromeWindowPosition, 1000);
            }

            // Auto minimize/restore
            if (vm.optionsSettings.autoMinimize) {
                $interval(vm.checkChromeWindowState, 1000);
            }

            vm.loadTabs();

            // Revert Main Chrome Window to initial width when closing Side Tabs
            window.onbeforeunload = vm.onExit;

            vm.initChromeEvents();

        }

        function initChromeEvents() {

            chrome.windows.onFocusChanged.addListener(function (windowId) {
                vm.windowChanged(windowId)
            });
            chrome.windows.onCreated.addListener(function (tab) {
                vm.windowChanged();
            });
            chrome.windows.onRemoved.addListener(function (windowId) {
                vm.windowChanged();
            });
            chrome.tabs.onCreated.addListener(function (param) {
                vm.loadTabs();
            });
            chrome.tabs.onActivated.addListener(function (activeInfo) {
                vm.loadTabs();
            });
            chrome.tabs.onHighlighted.addListener(function (highlightInfo) {
                vm.loadTabs(function () {
                    if (highlightInfo.tabIds.length == 1) {
                        scrollTabIntoView({id: highlightInfo.tabIds[0]});
                    }
                });
            });
            chrome.tabs.onAttached.addListener(function (tabId, attachInfo) {
                vm.loadTabs();
            });
            chrome.tabs.onDetached.addListener(function (tabId, detachInfo) {
                vm.loadTabs();
            });
            chrome.tabs.onMoved.addListener(function (tabId, moveInfo) {
                vm.loadTabs();
            });
            chrome.tabs.onRemoved.addListener(function (tabId, removeInfo) {
                vm.loadTabs();
            });
            chrome.tabs.onUpdated.addListener(function (tabId, changeInfo, tab) {
                vm.loadTabs();
            });

        }


        function initPreferences() {
            chrome.storage.sync.get(['tab', 'optionsSettings', 'sideTabsSettings'], function (items) {

                if (!items.tab) {
                    items.tab = {
                        height: 25,
                        width: 250,
                        fontSize: 11,
                        textColor: '#000000',

                        backgroundColor: {
                            top: '#DBD7D9',
                            bottom: '#C4C3C5'
                        },

                        hoverColor: {
                            top: '#F5F5F5',
                            bottom: '#E6E6E6'
                        },

                        highlightColor: '#F5F5F5',
                        borderColor: '#A9A9A9'
                    }
                } else {
                    items.tab = JSON.parse(items.tab);
                }

                if (!items.optionsSettings) {
                    items.optionsSettings = {
                        openSide: 'left',
                        backgroundColor: '#D4D4D4',
                        singleInstance: true,
                        autoAdjustWidth: true,
                        autoMinimize: true
                    }
                } else {
                    items.optionsSettings = JSON.parse(items.optionsSettings);
                }

                vm.optionsSettings = items.optionsSettings;

                if (!items.sideTabsSettings) {
                    items.sideTabsSettings = {
                        tabWidth: 250,
                        showTabsForSelectedWindow: false,
                        showTabsGroupedPerWindow: false,
                        showPinnedTabs: false
                    }
                } else {
                    items.sideTabsSettings = JSON.parse(items.sideTabsSettings);
                }

                vm.settings = items.sideTabsSettings;

                var style = angular.element('<style>')
                    .prop('type', 'text/css')
                    .html('\
                        body {\
                            font-size: ' + items.tab.fontSize + 'px;\
                            background-color: ' + items.optionsSettings.backgroundColor + ';\
                         }\
                        .tab {\
                            color: ' + items.tab.textColor + ';\
                            border-color: ' + items.tab.borderColor + '; \
                            background: linear-gradient(' + items.tab.backgroundColor.top + ', ' + items.tab.backgroundColor.bottom + ');\
                            height: ' + items.tab.height + 'px;\
                        }\
                        .tab.highlighted {\
                            background: ' + items.tab.highlightColor + '\
                        }\
                        .tab:hover {\
                            background: linear-gradient(' + items.tab.hoverColor.top + ', ' + items.tab.hoverColor.bottom + ');\
                        }\
                        .separator {\
                            background: + ' + items.tab.borderColor + ';\
                        }\
                        .window-separator {\
                            color: ' + items.tab.borderColor + ' \
                        }\
                        ');

                angular.element(document.querySelector('head')).append(style);


                vm.initSideTabs();

            })
        }

        function savePreferences() {

            chrome.storage.sync.set({
                sideTabsSettings: JSON.stringify(vm.settings)
            })

        }

        function clickCreateNewTab() {

            vm.isNewTabFromSideTabs = true;

            var windowId = vm.data.idLastChromeWindow ? vm.data.idLastChromeWindow : chrome.windows.WINDOW_ID_CURRENT;

            chrome.tabs.create({
                'windowId': windowId
            })

        }

        function clickTab(tab) {

            try {

                chrome.windows.get(tab.windowId, function (window) {

                    if (window.state == 'minimized' || vm.data.idLastChromeWindow != tab.windowId) {
                        chrome.windows.update(tab.windowId, {'focused': true}, function () {
                            chrome.tabs.update(tab.id, {'active': true}, function () {
                                chrome.windows.update(vm.data.idSideTabsWindow, {'focused': true}, function () {
                                    vm.loadTabs(function () {
                                        scrollTabIntoView(tab);
                                    });
                                });
                            });
                        })
                    }
                    else {
                        chrome.tabs.update(tab.id, {'active': true}, function () {
                            chrome.windows.update(vm.data.idSideTabsWindow, {'focused': true}, function () {
                                vm.loadTabs(function () {
                                    scrollTabIntoView(tab);
                                });
                            });
                        });
                    }
                })

            }
            catch (ex) {
                console.log(ex);
            }

        }

        function clickCloseTab(tab) {

            chrome.tabs.remove(tab.id, function () {
                vm.loadTabs();
            })

        }

        function clickShowSettings() {
            vm.settings.showSettings = !vm.settings.showSettings;
            vm.savePreferences();
        }

        function clickShowSearch() {
            vm.settings.showSearch = !vm.settings.showSearch;
            vm.savePreferences();

            if (!vm.settings.showSearch) {
                vm.data.search = '';
                return;
            }

            setTimeout(function () {
                document.getElementById('search').focus();
            }, 50);
        }

        function clickShowTabsForSelectedWindow() {
            vm.settings.showTabsForSelectedWindow = !vm.settings.showTabsForSelectedWindow;
            vm.savePreferences();
            vm.loadTabs();
        }

        function clickShowTabsGroupedPerWindow() {
            vm.settings.showTabsGroupedPerWindow = !vm.settings.showTabsGroupedPerWindow;
            vm.savePreferences();
        }

        function clickShowPinnedTabs() {
            vm.settings.showPinnedTabs = !vm.settings.showPinnedTabs;
            vm.savePreferences();
        }

        function clickMoreSettings() {
            chrome.runtime.openOptionsPage();
        }

        function loadTabs(callback) {

            chrome.windows.getAll({"populate": true}, function (windows) {

                // Reset number of tabs to 0;
                vm.data.numberOfTabs = 0;

                var filteredWindows = [];

                for (var i = 0; i < windows.length; i++) {
                    if (windows[i].type == "normal") {

                        if (vm.settings.showTabsForSelectedWindow) {
                            if (windows[i].id === vm.data.idLastChromeWindow) {
                                filteredWindows.push(windows[i]);
                            }
                        } else {
                            filteredWindows.push(windows[i]);
                        }

                        for (var j = 0; j < windows[i].tabs.length; j++) {
                            vm.data.numberOfTabs++;
                        }
                    }

                    /*
                     if (windows[i].id == vm.data.idLastChromeWindow) {
                     for (var j = 0; j < windows[i].tabs.length; j++) {
                     if (windows[i].tabs[j].highlighted) {
                     scrollTabIntoView(windows[i].tabs[j]);
                     }
                     }
                     }
                     */
                }

                vm.data.windows = filteredWindows;

                if (!$scope.$$phase) {
                    $scope.$digest();
                }


                /*
                for (var i = 0; i < vm.data.windows.length; i++) {
                    if (vm.data.windows[i].id == vm.data.idLastChromeWindow) {
                        for (var j = 0; j < vm.data.windows[i].tabs.length; j++) {
                            if (vm.data.windows[i].tabs[j].highlighted) {
                            }
                        }
                    }
                }
                */

                if (callback) {
                    callback();
                }

            });

        }

        function isScrolledIntoView(tab) {

            if (!tab) return;

            var tabsWrapper = document.querySelector('.tabs-wrapper');

            return ((tab.offsetTop + tab.offsetHeight <= (tabsWrapper.offsetHeight + tabsWrapper.scrollTop) * 0.9) &&
            (tab.offsetTop >= (tabsWrapper.scrollTop + tabsWrapper.offsetTop) * 1.1));

        }

        function scrollTabIntoView(selectedTab) {

            var tabs = document.querySelectorAll('.tab');

            for (var i = 0; i < tabs.length; i++) {
                var tab = angular.element(tabs[i]).scope().tab;
                if (!tab) continue;

                if (tab.id == selectedTab.id) {
                    var tabDiv = tabs[i];
                    break;
                }
            }

            if (vm.isScrolledIntoView(tabDiv) == false) {

                var tabsWrapper = document.querySelector('.tabs-wrapper');
                tabsWrapper.scrollTop = tabDiv.offsetTop - tabsWrapper.offsetHeight / 2;

            }
        }

        function checkChromeWindowPosition() {

            chrome.windows.get(vm.mainChromeWindow.id, function (window) {

                if (vm.mainChromeWindow.left != window.left ||
                    vm.mainChromeWindow.width != window.width) {

                    vm.mainChromeWindow.left = window.left;
                    vm.mainChromeWindow.width = window.width;

                    if (vm.optionsSettings.openSide === 'left') {

                        vm.settings.tabWidth = window.left;

                        chrome.windows.update(vm.data.idSideTabsWindow, {
                            width: vm.settings.tabWidth,
                            left: 0,
                            top: 0
                        });

                    } else {

                        vm.settings.tabWidth = screen.width - window.width;

                        chrome.windows.update(vm.data.idSideTabsWindow, {
                            left: window.width,
                            top: 0,
                            width: vm.settings.tabWidth
                        });

                    }

                    if (vm.settings.tabWidth > 30 && vm.settings.tabWidth < window.width - 30) {
                        savePreferences();
                    }

                }

            });

        }

        function checkChromeWindowState() {

            chrome.windows.get(vm.mainChromeWindow.id, function (window) {

                if (vm.mainChromeWindow.state != window.state) {

                    vm.mainChromeWindow.state = window.state;

                    if (window.state == 'minimized') {
                        chrome.windows.update(vm.data.idSideTabsWindow, {
                            state: 'minimized'
                        });
                    } else {
                        chrome.windows.update(vm.data.idSideTabsWindow, {
                            state: 'normal'
                        });
                    }

                }

            });

        }

        function windowChanged(windowId) {

            if (windowId != vm.data.idSideTabsWindow && windowId >= 0) {
                vm.data.idLastChromeWindow = windowId;
                vm.loadTabs();
            }

        }

        function onExit() {

            chrome.windows.update(vm.mainChromeWindow.id, {
                width: screen.width,
                left: 0
            });

        }

        vm.init();

    }

})();